File: internal/cli/tunnel.go
package cli

import (
	"errors"
	"fmt"
	"os"
	"text/tabwriter"

	"github.com/spf13/cobra"
	"github.com/starfield17/rhost/internal/errs"
	"github.com/starfield17/rhost/internal/output"
	"github.com/starfield17/rhost/internal/transport/openssh"
)

// newTunnelCmd manages port forwards held by dedicated OpenSSH masters.
//
// The three subcommands are deliberately thin: every decision about what a
// forward may bind, and whether one is alive, belongs to the transport and to
// OpenSSH. What this adds is the agent-facing contract — an ID to keep, a status
// to branch on, and a JSON document for each.
func newTunnelCmd() *cobra.Command {
	group := newGroup("tunnel", "Manage persistent OpenSSH tunnels",
		"Tunnels are dedicated OpenSSH masters: they outlive the CLI process that\n"+
			"opened them, and nothing here restarts them after a reboot. The ID printed\n"+
			"by `open` is what `close` takes; `list` redisovers them on this machine.")

	var kind, listen, destination string
	var expose bool

	open := &cobra.Command{
		Use:   "open <host>",
		Short: "Open a persistent port forward",
		Long: `Open one forward on its own OpenSSH master.

kinds:
  local    --listen localhost:8080 --destination localhost:8000
           reach a service on the *remote* side through this machine
  reverse  --listen localhost:9000 --destination localhost:3000
           reach a service on *this* machine through the remote one
  socks    --listen localhost:1080
           a SOCKS5 proxy through the remote host

The bind defaults to loopback. Anything else lets other machines in through the
forward and has to be asked for with --allow-exposure. The remote sshd has its own
say: a denied reverse bind is reported as OpenSSH said it, never worked around.`,
		Args: cobra.ExactArgs(1),
		RunE: func(c *cobra.Command, args []string) error {
			audit := startAudit("tunnel.open", args[0])
			if aerr := validateTunnelOpen(kind, listen, destination, expose); aerr != nil {
				audit.fail(aerr)
				emitFailure("tunnel.open", args[0], aerr)
				return nil
			}
			t, err := tunnelClient().OpenTunnel(c.Context(), args[0], kind, listen, destination, expose)
			if err != nil {
				// Retryable, because the common causes here are transport-shaped (a
				// dropped master, a host that is briefly down). A bind sshd *refused*
				// will refuse again, so the message has to carry OpenSSH's own words —
				// which it does, verbatim (§33).
				aerr := errs.Wrap(errs.TunnelFailed, err.Error(), true, err)
				audit.fail(aerr)
				emitFailure("tunnel.open", args[0], aerr)
				return nil
			}
			audit.succeed("", kind+" "+listen+" -> "+destination, nil)
			emitTunnel("tunnel.open", args[0], t)
			return nil
		},
	}
	open.Flags().StringVar(&kind, "kind", "local", "local, reverse or socks")
	open.Flags().StringVar(&listen, "listen", "localhost:8080", "address:port to bind")
	open.Flags().StringVar(&destination, "destination", "",
		"host:port to forward to (not used for socks)")
	open.Flags().BoolVar(&expose, "allow-exposure", false,
		"permit a bind outside loopback (destructive: other machines can reach the target)")

	list := &cobra.Command{
		Use:   "list",
		Short: "List tunnels and check each master",
		Long: `List the tunnels this machine has records for, and ask each one's OpenSSH
master whether it is still up. A record whose master has gone is reported as
"stale" and left in place — it is evidence, not rhost's to discard.

An alive master is not the same as a healthy service: this checks the forward
exists, not that anything is answering on the other end.`,
		Args: cobra.NoArgs,
		RunE: func(c *cobra.Command, _ []string) error {
			rows, err := tunnelClient().ListTunnels(c.Context())
			if err != nil {
				emitFailure("tunnel.list", "", tunnelErr(err))
				return nil
			}
			emitTunnels(rows)
			return nil
		},
	}

	closeCmd := &cobra.Command{
		Use:   "close <id>",
		Short: "Close one tunnel and its own master",
		Long: `Close exactly one tunnel: its dedicated master is asked to exit and its
record is removed. The shared connection used by exec, session and job traffic is
untouched, and other tunnels keep running.`,
		Args: cobra.ExactArgs(1),
		RunE: func(c *cobra.Command, args []string) error {
			audit := startAudit("tunnel.close", "")
			err := tunnelClient().CloseTunnel(c.Context(), args[0])
			if err != nil {
				aerr := tunnelErr(err)
				audit.fail(aerr)
				emitFailure("tunnel.close", "", aerr)
				return nil
			}
			audit.succeed("", args[0], nil)
			if jsonFlag {
				writeEnvelope(output.Success("tunnel.close", "", map[string]interface{}{
					"id": args[0], "closed": true,
				}))
			} else {
				fmt.Fprintf(os.Stderr, "closed %s\n", args[0])
			}
			return nil
		},
	}

	group.AddCommand(open, list, closeCmd)
	return group
}

// tunnelClient is the OpenSSH client these commands talk through. A tunnel's
// master is chosen by the same options every other command uses, so the only
// difference between `tunnel open` and `exec` is which socket they bind.
func tunnelClient() *openssh.Client {
	return openssh.New(openssh.DefaultConfig())
}

// tunnelErr maps a transport failure onto the taxonomy. A missing record is not
// the same problem as a refused forward, and an agent that wants to clean up
// needs to be able to tell "already gone" from "could not be stopped".
func tunnelErr(err error) *errs.Error {
	if errors.Is(err, openssh.ErrInvalidTunnel) {
		return errs.Wrap(errs.ConfigInvalid, err.Error(), false, err)
	}
	if errors.Is(err, openssh.ErrNoTunnel) {
		return errs.Wrap(errs.TunnelNotFound, err.Error(), false, err)
	}
	return errs.Wrap(errs.TunnelFailed, err.Error(), true, err)
}

func validateTunnelOpen(kind, listen, destination string, expose bool) *errs.Error {
	if err := openssh.ValidateTunnel(kind, listen, destination, expose); err != nil {
		return tunnelErr(err)
	}
	return nil
}

// emitTunnel renders one tunnel; `open` returns the record it just created.
func emitTunnel(operation, host string, t openssh.Tunnel) {
	if jsonFlag {
		writeEnvelope(output.Success(operation, host, t))
		return
	}
	fmt.Printf("%s  %s  %s  %s  %s\n", t.ID, t.Status, t.Kind, t.Listen, t.Destination)
	fmt.Fprintf(os.Stderr, "keep this id: rhost tunnel close %s\n", t.ID)
}

func emitTunnels(rows []openssh.Tunnel) {
	if jsonFlag {
		writeEnvelope(output.Success("tunnel.list", "", map[string]interface{}{"tunnels": rows}))
		return
	}
	if len(rows) == 0 {
		fmt.Fprintln(os.Stderr, "no tunnels")
		return
	}
	w := tabwriter.NewWriter(os.Stdout, 0, 2, 2, ' ', 0)
	fmt.Fprintln(w, "ID\tSTATUS\tKIND\tLISTEN\tDESTINATION\tHOST")
	for _, t := range rows {
		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\n",
			t.ID, t.Status, t.Kind, t.Listen, dash(t.Destination), t.Host)
	}
	_ = w.Flush()
}

File: src/cli/tunnel.rs
//! The `tunnel` command surface: grammar, running and the mapping of a tunnel
//! fault onto the error taxonomy.
//!
//! The three subcommands are thin on purpose. Every decision about what a forward
//! may bind, and whether one is alive, belongs to `crate::tunnel` and to OpenSSH;
//! what this adds is the agent-facing contract — an id to keep, a status to
//! branch on, and one document per call (AGENTS.md §6).

use super::audit::Timer;
use super::console::{Failure, Sink};
use super::grammar::{FlagSpec, JSON, PLAIN_FLAGS, help_or_error, parse, usage_error};
use super::usage::leaf_or_group;
use super::{Command, Help, Scope};
use crate::output;
use crate::transport::Client;
use crate::tunnel::{self, Fault};

/// Every tunnel operation the parser can hand to [`run`].
pub enum Tunnel {
    Open {
        host: String,
        kind: String,
        listen: String,
        destination: String,
        expose: bool,
    },
    List,
    Close {
        id: String,
    },
}

/// The flags `open` takes. The group accepts them so the leaf can be found, and
/// every other leaf then parses the same argv again with its own table, which is
/// what refuses a flag that belongs to a sibling operation.
static TUNNEL_FLAGS: &[FlagSpec] = &[
    JSON,
    FlagSpec::value("kind", false, "forward kind (local or remote)"),
    FlagSpec::value("listen", false, "address to listen on"),
    FlagSpec::value("destination", false, "forward destination host:port"),
    FlagSpec::long("allow-exposure", "permit binding a non-loopback address"),
];

/// The flags each leaf accepts, so its `--help` page and its parse read the same
/// table. `list` and `close` take operands only.
pub(crate) fn leaf_flags(leaf: &str) -> &'static [FlagSpec] {
    if leaf == "open" {
        TUNNEL_FLAGS
    } else {
        PLAIN_FLAGS
    }
}

/// Defaults that match the reference: the common case is a local forward out of
/// loopback, and a caller who wants something else says so.
const DEFAULT_KIND: &str = "local";
const DEFAULT_LISTEN: &str = "localhost:8080";

pub(crate) fn command(argv: &[String], json: bool) -> Command {
    let scanned = match parse(argv, TUNNEL_FLAGS) {
        Ok(parsed) => parsed,
        Err(message) => return usage_error(Scope::Tunnel, message),
    };
    let leaf = scanned.operand(0).map(str::to_string);
    if scanned.bool_flag("help") {
        return help_or_error(
            Scope::Tunnel,
            match &leaf {
                Some(leaf) => leaf_or_group(Scope::Tunnel, leaf),
                None => Help::Group(Scope::Tunnel),
            },
            json,
        );
    }
    let Some(leaf) = leaf else {
        return usage_error(Scope::Tunnel, "rhost tunnel needs a subcommand".to_string());
    };
    let parsed = match parse(argv, leaf_flags(&leaf)) {
        Ok(parsed) => parsed,
        Err(message) => return usage_error(Scope::Tunnel, message),
    };
    let operands = &parsed.operands[1.min(parsed.operands.len())..];
    match leaf.as_str() {
        "open" => {
            if let Err(message) = exact(&leaf, operands, 1, "<host>") {
                return usage_error(Scope::Tunnel, message);
            }
            let flag =
                |name: &str, default: &str| parsed.value(name).unwrap_or(default).to_string();
            Command::Tunnel(Tunnel::Open {
                host: operands[0].clone(),
                kind: flag("kind", DEFAULT_KIND),
                listen: flag("listen", DEFAULT_LISTEN),
                destination: flag("destination", ""),
                expose: parsed.bool_flag("allow-exposure"),
            })
        }
        "list" => {
            if let Err(message) = exact(&leaf, operands, 0, "") {
                return usage_error(Scope::Tunnel, message);
            }
            Command::Tunnel(Tunnel::List)
        }
        "close" => {
            if let Err(message) = exact(&leaf, operands, 1, "<id>") {
                return usage_error(Scope::Tunnel, message);
            }
            Command::Tunnel(Tunnel::Close {
                id: operands[0].clone(),
            })
        }
        other => usage_error(
            Scope::Tunnel,
            format!("unknown command {other} for \"rhost tunnel\""),
        ),
    }
}

fn exact(leaf: &str, operands: &[String], count: usize, names: &str) -> Result<(), String> {
    if operands.len() == count {
        return Ok(());
    }
    let wanted = if names.is_empty() {
        String::new()
    } else {
        format!(" {names}")
    };
    Err(format!(
        "rhost tunnel {leaf} accepts {count} arg(s){wanted}, received {}",
        operands.len()
    ))
}

/// Runs one tunnel operation, delivering the envelope or the human view, and
/// mapping a fault exactly once so a code never has to be repeated.
pub(crate) fn run(sink: &mut Sink, client: &Client, command: Tunnel, json: bool) -> u8 {
    match command {
        Tunnel::Open {
            host,
            kind,
            listen,
            destination,
            expose,
        } => {
            let timer = Timer::start("tunnel.open", &host);
            let request = match tunnel::Request::new(&host, &kind, &listen, &destination, expose) {
                Ok(request) => request,
                Err(fault) => {
                    let failure = failure("tunnel.open", &host, fault);
                    timer.failed(failure.code());
                    return failure.deliver(sink, json);
                }
            };
            match tunnel::open(client, &request) {
                Ok(opened) => {
                    timer.succeeded("", &format!("{kind} {listen} -> {destination}"), None);
                    if json {
                        sink.envelope(&output::tunnel("tunnel.open", &host, &opened));
                    } else {
                        super::render::tunnel(sink, &opened);
                    }
                    0
                }
                Err(fault) => {
                    let failure = failure("tunnel.open", &host, fault);
                    timer.failed(failure.code());
                    failure.deliver(sink, json)
                }
            }
        }
        Tunnel::List => match tunnel::list(client) {
            Ok(rows) => {
                if json {
                    sink.envelope(&output::tunnels(&rows));
                } else {
                    super::render::tunnels(sink, &rows);
                }
                0
            }
            Err(fault) => failure("tunnel.list", "", fault).deliver(sink, json),
        },
        Tunnel::Close { id } => {
            let timer = Timer::start("tunnel.close", "");
            match tunnel::close(client, &id) {
                Ok(closed) => {
                    timer.succeeded("", &closed, None);
                    if json {
                        sink.envelope(&output::tunnel_closed(&closed));
                    } else {
                        super::console::warn(&format!("closed {closed}"));
                    }
                    0
                }
                Err(fault) => {
                    let failure = failure("tunnel.close", "", fault);
                    timer.failed(failure.code());
                    failure.deliver(sink, json)
                }
            }
        }
    }
}

/// A missing record is not the same problem as a refused forward: an agent that
/// wants to clean up has to be able to tell "already gone" from "could not be
/// stopped", and only the transport-shaped failure is worth retrying.
fn failure(operation: &'static str, host: &str, fault: Fault) -> Failure {
    match fault {
        Fault::Invalid(message) => Failure::new(operation, host, "CONFIG_INVALID", message),
        Fault::NotFound(message) => Failure::new(operation, host, "TUNNEL_NOT_FOUND", message),
        Fault::Transport(message) => {
            Failure::new(operation, host, "TUNNEL_FAILED", message).retryable()
        }
    }
}
